pkg/gentrait/securityevent/model.go

```diff
@@ -0,0 +1,166 @@
+	allSecurityEvents []*gen.SecurityEvent
+	genId             int
+
+	lastSecurityEvent *resource.Value // of *gen.SecurityEvent
+}
+
+func NewModel(opts ...resource.Option) *Model {
+	defaultOpts := []resource.Option{resource.WithInitialValue(&gen.SecurityEvent{})}
+	opts = append(defaultOpts, opts...)
+
+	m := &Model{
+		lastSecurityEvent: resource.NewValue(opts...),
+	}
+
+	// let's add some events to start with so we can test the list method without waiting
+	startTime := time.Now().Add(-100 * time.Minute)
+	for i := 0; i < 100; i++ {
+		_, _ = m.GenerateSecurityEvent(timestamppb.New(startTime))
+		startTime = startTime.Add(time.Minute)
+	}
+
+	return m
+}
+
+// AddSecurityEvent manually add a security event to the model
+func (m *Model) AddSecurityEvent(se *gen.SecurityEvent, opts ...resource.WriteOption) (*gen.SecurityEvent, error) {
+	v, err := m.lastSecurityEvent.Set(se, opts...)
+	if err != nil {
+		return nil, err
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.allSecurityEvents = append(m.allSecurityEvents, se)
+	return v.(*gen.SecurityEvent), nil
+}
+
+// GenerateSecurityEvent generates a security event and adds it to the model
+func (m *Model) GenerateSecurityEvent(time *timestamppb.Timestamp) (*gen.SecurityEvent, error) {
+	m.mu.Lock()
+	se := &gen.SecurityEvent{
+		SecurityEventTime: time,
+		Description:       "Generated security event",
+		Id:                strconv.Itoa(m.genId),
+		Source: &gen.SecurityEvent_Source{
+			Subsystem: subSystems[m.genId%2],
+		},
+		EventType: gen.SecurityEvent_EventType(m.genId % 22),
+	}
+	m.genId++
+	m.mu.Unlock()
+	return m.AddSecurityEvent(se)
+}
+
+func (m *Model) ListSecurityEvents(req *gen.ListSecurityEventsRequest) (*gen.ListSecurityEventsResponse, error) {
+	// page token is just the index of where we left off (if any)
+	// this works with the current basic implementation because we only support a list of all events without filtering/sorting
+	// and the events are stored in ascending chronological order. If this either of these things change, this will need to be rethought
+	pageToken := req.GetPageToken()
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	startIndex := len(m.allSecurityEvents)
+	if pageToken != "" {
+		_, err := strconv.Atoi(req.GetPageToken())
+		if err != nil {
+			return nil, err
+		}
+		startIndex, _ = strconv.Atoi(pageToken)
+	}
+
+	count := req.PageSize
+	if count == 0 {
+		count = 50
+	} else if count > 1000 {
+		count = 1000
+	}
+
+	resp := &gen.ListSecurityEventsResponse{}
+
+	// reverse to retrieve the latest events first
+	for i := startIndex - 1; i >= 0; i-- {
+		resp.SecurityEvents = append(resp.SecurityEvents, m.allSecurityEvents[i])
+		if len(resp.SecurityEvents) >= int(count) {
+			resp.NextPageToken = strconv.Itoa(i - 1)
+			break
+		}
+	}
+	resp.TotalSize = int32(len(m.allSecurityEvents))
+	return resp, nil
+}
+
+func (m *Model) PullSecurityEventsWrapper(request *gen.PullSecurityEventsRequest, server gen.SecurityEventApi_PullSecurityEventsServer) error {
```

Comment:
Other models that need to do this have function that return both current data and a stream like resource.Collection does.

If you want to keep this func here then it shouldn't be exported as its not really something people who just have a model can use.